hathor/indexes/height_index.py

```diff
@@ -0,0 +1,94 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import List, NamedTuple, Optional, Tuple
+
+from hathor.transaction import Block
+from hathor.transaction.genesis import BLOCK_GENESIS
+from hathor.util import not_none
+
+
+class IndexEntry(NamedTuple):
+    """Helper named tuple that implementations can use."""
+    hash: bytes
+    timestamp: int
+
+
+BLOCK_GENESIS_ENTRY: IndexEntry = IndexEntry(not_none(BLOCK_GENESIS.hash), BLOCK_GENESIS.timestamp)
+
+
+class _AddToIndexItem(NamedTuple):
+    height: int
+    hash: bytes
+    timestamp: int
+
+
+class HeightIndex(ABC):
+    """Store the block hash for each given height
+    """
+
+    @abstractmethod
+    def add_new(self, height: int, block_hash: bytes, timestamp: int) -> None:
+        """Add a new block to the height index that **MUST NOT** result in a re-org"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def add_reorg(self, height: int, block_hash: bytes, timestamp: int) -> None:
+        """Add a new block to the height index that **MIGHT** result in a re-org"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def get(self, height: int) -> Optional[bytes]:
+        """ Return the block hash for the given height, or None if it is not set.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_tip(self) -> bytes:
+        """ Return the best block hash, it returns the genesis when there is no other block
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_height_tip(self) -> Tuple[int, bytes]:
+        """ Return the best block height and hash, it returns the genesis when there is no other block
+        """
+        raise NotImplementedError
+
+    def update_new_chain(self, height: int, block: Block) -> None:
+        """ When we have a new winner chain we must update all the height index
+            until the first height with a common block
+        """
+        assert self.get(height) != block.hash
+
+        block_height = height
+        side_chain_block = block
+        add_to_index: List[_AddToIndexItem] = []
+        while self.get(block_height) != side_chain_block.hash:
+            add_to_index.append(
+                _AddToIndexItem(block_height, not_none(side_chain_block.hash), side_chain_block.timestamp)
+            )
+
+            side_chain_block = side_chain_block.get_block_parent()
+            new_block_height = side_chain_block.get_metadata().height
+            assert new_block_height + 1 == block_height
+            block_height = new_block_height
+
+        # Reverse the data because I was adding in the array from the highest block
+        reversed_add_to_index = reversed(add_to_index)
+
+        for item in reversed_add_to_index:
+            # Add it to the index
+            self.add_reorg(item.height, item.hash, item.timestamp)
```

Comment:
The logic here is definitely simpler, but I wonder if it will have any impact on performance. The deletion of the old heights will occur as a batch, but adding new indexes will be done one by one. Anyway, this is probably a corner case and we shouldn't worry about it. Maybe even there won't be much impact, as the reorgs are likely only a few blocks. Long reorgs should be very rare.

Reply:
This logic is not new at all, it's exactly the same as before, just on a different place. While I agree we could have done some optimization before to avoid adding one-by-one, we didn't do this before. It could be possible on the memory index, but I don't see how to do that on the rocksdb index. Anyway, since we aren't changing anything, it seemed fine for me. But we could take the opportunity to rethink if it makes sense.

Reply:
I was actually thinking for the rocks db index. It seems they have batch operations. Anyway, I don't think it's worth doing it. I just brought it up so you'd also think if it was worth it.

Reply:
Ah, got it. I think this could be just a matter of refactor so this operation would be done in a batch instead. It shouldn't be too hard. I think there is a good advantage, not in the sense of performance, it probably won't have any noticeable impact, but it would be slightly better in case of a crash.

Reply:
I thought it would be simpler but since this is on the base class it would need more abstractions. I'll leave it as is right now.